tests: arch: add test case of arch_nop() interface

enjiamai opened the pull request:
Add a test case to test arch interface arch_nop(), the main focus here
is for coverage of the code. arch_nop() is a special implementation and
it will behave differently on different platforms. By the way, this also
measures how many cycles it spends for platforms that support it.

Signed-off-by: Enjia Mai <enjiax.mai@intel.com>

Commit:
tests: arch: add a test case for testing arch_nop() interface

Add a test case to test arch interface arch_nop(), the main focus here
is for coverage of the code. arch_nop() is a special implementation
and it will behave differently on different platforms. By the way, this
also measures how many cycles it spends for platforms that support it.

Signed-off-by: Enjia Mai <enjiax.mai@intel.com>

LeiW000 commented on tests/kernel/arch_interface/misc/src/main.c:
```diff
@@ -0,0 +1,64 @@
+#define TEST_ROUND	(1 << SHIFT_BITS)
+
+/**
+ * @brief Test how much time arch_nop() costs
```
It better to replace "costs" with "takes".

enjiamai replied:
Correct it, thanks!

LeiW000 commented on tests/kernel/arch_interface/misc/src/main.c:
```diff
@@ -0,0 +1,64 @@
+	t_get_time = t_after - t_before;
+
+	/* if k_cycle_get_32() cost no time means test not works */
+	if (t_get_time == 0) {
```
Is it to check if a platform has the capability to run the test? If yes, please add comments for it.

enjiamai replied:
Yes, it is. I refined the comments as you mentioned, thanks!

LeiW000 commented on tests/kernel/arch_interface/misc/src/main.c:
```diff
@@ -0,0 +1,64 @@
+
+	t_after = k_cycle_get_32();
+
+	/* calculate cost time of arch_nop() */
```
We can say "Calculate delta time of arch_nop()".

enjiamai replied:
Got it fix, thanks!

enjiamai commented:
@LeiW000 I have addressed the comment you give, could you please review it again, thanks!

andyross reviewed (changes requested):
Adding a whole new test just to get coverage on a function that **literally does nothing** seems maybe a little heavyweight to me. :)

Can't this be a test case is tests/kernel/common or somewhere similar?

enjiamai commented:
> Adding a whole new test just to get coverage on a function that **literally does nothing** seems maybe a little heavyweight to me. :)
> 
> Can't this be a test case is tests/kernel/common or somewhere similar?

Hi @andyross , yes, add a directory just for such a test case is not a good idea. My thinking here is trying to add a new test suite for some more coming arch interface testing I'll working now (mainly target to improve code coverage). Do you think add this test suite into a new directory tests/arch/common is acceptable?  Or just try to add this sort of test into current existing test suite is recommended?  such as tests/kernel/common you mentioned.

andyross commented:
Generally we want to avoid adding whole new tests, they have to be built individually and are a significant burden on CI when added.   "3-4 seconds to build * a dozen or more platforms / always_increasing_CI_parallelism" isn't fun math.  But I agree an arch-layer test would be a good idea.  Maybe wait until we have more than arch_nop to add it though?  Or steal the irq_offload and spinlock test and some others to combine them to start it with?

enjiamai commented:
> Generally we want to avoid adding whole new tests, they have to be built individually and are a significant burden on CI when added. "3-4 seconds to build * a dozen or more platforms / always_increasing_CI_parallelism" isn't fun math. But I agree an arch-layer test would be a good idea. Maybe wait until we have more than arch_nop to add it though? Or steal the irq_offload and spinlock test and some others to combine them to start it with?

Yes, got your point and concern. I'll add them to some other existing test case such as you mentioned, then move them until we have sufficient test cases to form the arch layer test suite, thank you so much!

nashif reviewed (changes requested):
agree with Andy, not sure what we are testing here and it does not make sense to have one test to get coverage of an API.

enjiamai commented:
> agree with Andy, not sure what we are testing here and it does not make sense to have one test to get coverage of an API.

Hi, @nashif , I did not come up with a better way to test this arch_nop() function, it does nothing but consume CPU cycle, so I try to call it for showing that arch_nop() takes different cycle in different platform. I'll consider other scenario, thanks!

jharris-intel commented:
As an FYI on the ARM side of things:

`arch_nop` redirects to `nop`, on both AArch32 and AArch64.

However, the ARMv8-A ARM states the following:

> No Operation does nothing, other than advance the value of the program counter by 4. This instruction can be used
for instruction alignment purposes.
> Note
> The timing effects of including a NOP instruction in a program are not guaranteed. It can increase execution time,
leave it unchanged, or even reduce it. Therefore, NOP instructions are not suitable for timing loops.

(Note: this is also true for ARMv7-A/R; ARMv7-M does not have this wording.)

So if this is indeed the desired purpose of `arch_nop`, we cannot use `nop` for the implementation of `arch_nop` on generic Cortex A or R cores.

andyross commented:
Really the root issue here is that there is no desired purpose of `arch_nop()`; it's dead code.  We spec it, architectures define it, but nothing ever calls it.

enjiamai commented:
@jharris-intel Thanks for your information and reminder~ understand why it take zero cycle per call in some platform.

enjiamai commented:
> Really the root issue here is that there is no desired purpose of `arch_nop()`; it's dead code. We spec it, architectures define it, but nothing ever calls it.

So @andyross , I guess I should close this PR and left arch_nop() there and not trying to cover it this way unless a better usage comes out, right?

andyross commented:
I don't see anything wrong with having a coverage test in place for arch_nop() even if we don't use it.  We probably don't want to remove it from the arch layer, even if we don't actually want it right now.   I just didn't like this being a standalone test.

jharris-intel commented:
> Really the root issue here is that there is no desired purpose of arch_nop(); it's dead code. We spec it, architectures define it, but nothing ever calls it.

Potential uses of `arch_nop`:

1. Code alignment -> although in this case it's much more likely the _compiler_ doing so (or you're in an assembly file, in which case you're not calling `arch_nop` anyway). And this would require that `arch_nop` be `ALWAYS_INLINE`.
2. Giving you a guaranteed place to put a breakpoint / trace trigger / etc. when debugging. This is my main usage of `arch_nop`; it inherently is generally debugging code removed before actually pushing.
3. Giving you a guaranteed place to put a patchpoint. E.g. ARMv7 allows `nop` (and a few other instructions) to be modified concurrently with execution, but not most other instructions.
4. Delaying a few instructions, e.g. for tight timing loops on M-cores. 

It may be worth clarifying which of these it is intended for.

enjiamai commented:
@jharris-intel  Thanks so much for suggesting this scenario, I can try that~ it might take some time~

LeiW000 commented:
@enjiamai , it seems that you need to redesign the test case. Am I right?

enjiamai commented:
> @enjiamai , it seems that you need to redesign the test case. Am I right?

Yes, @steven, it needs to redesign. According to @jharris-intel mentioned, I think scenario 4 is easier to write a testcase on it.

enjiamai commented:
Hi @andyross @nashif , I did not come up with a better test case for this, so I make some change of it and move it to kernel.common, could you please help to review again, thanks!

enjiamai commented:
Hi @andyross, I think I have addressed your comments mentioned before, I put them in kernel/common and not be a standalone testsuite. could you please help to review it again? thank you!

andyross reviewed (approved):
I'd still feel better if we just killed this function, which isn't specified well enough to be useful.  But this is as good a test for it as I can imagine.

LeiW000 commented:
@nashif , I guess that it can be merged now. Could you help to merge it?
